### PR TITLE
Two bugs fixes

### PR DESCRIPTION
**Pull request by liulinC**
CA-356655: subject-add failed with ad.xenrt.citrite.net
    
    To support multiple trusted domain, the domain/server name needs
    to be provided during ldap search.
    Before this commit, the domain is provided, net command will find
    the best server to perform ldap search.
    However, in some scenario like lab, DNS failed to resovle the domain
    name as dns configuration.
    This commit fix the issue by lookup kdc and provide the IP address
    instead to skip the DSN resolve

CA-356674: Hostname present on AD server not as expected
    
    xenrt use advanced feature of PBIS to customize host dns name
    This commit expose dns-hsotname and netbios-name option to advanced
    users, and use default value if not provided.
    netbios-name should be limitted to 15 chars

### CODE REVIEW

File: ocaml/xapi/extauth_plugin_ADwinbind.ml [hunk 1]
```diff
@@ -554,7 +568,8 @@ let config_winbind_damon ~domain ~workgroup ~netbios_name =
       ; "winbind enum groups = no"
       ; "winbind enum users = no"
       ; Printf.sprintf "winbind cache time = %d" !Xapi_globs.winbind_cache_time
-      ; Printf.sprintf "machine password timeout = %d" !Xapi_globs.winbind_machine_pwd_timeout
+      ; Printf.sprintf "machine password timeout = %d"
```
Comment (liulinC) [inline ocaml/xapi/extauth_plugin_ADwinbind.ml:571]: Stange, `make format` gives me different format, 
is there any conguration change recently about the format?

**Comment by liulinC** [comment]
The unittest failed with  IDL error, Looks like I need to merge some change in idl repo, 
what should I do in this Koji world

```Error: This expression has type
40
         Cluster_interface.pems option ->
41
         (unit, Cluster_interface.error) Cluster_client.IDL.T.resultb
42
       but an expression was expected of type
43
         'a Cluster_client.IDL.T.box = 'a Idl.Make(Idl.IdM).T.box
44
make: *** [Makefile:13: build] Error 1
45
Error: Process completed with exit code 2.```

**Comment by lippirk** [comment]
> The unittest failed with IDL error, Looks like I need to merge some change in idl repo,
> what should I do in this Koji world
> 
> ```
> 40
>          Cluster_interface.pems option ->
> 41
>          (unit, Cluster_interface.error) Cluster_client.IDL.T.resultb
> 42
>        but an expression was expected of type
> 43
>          'a Cluster_client.IDL.T.box = 'a Idl.Make(Idl.IdM).T.box
> 44
> make: *** [Makefile:13: build] Error 1
> 45
> Error: Process completed with exit code 2.```
> ```



> The unittest failed with IDL error, Looks like I need to merge some change in idl repo,
> what should I do in this Koji world
> 
> ```
> 40
>          Cluster_interface.pems option ->
> 41
>          (unit, Cluster_interface.error) Cluster_client.IDL.T.resultb
> 42
>        but an expression was expected of type
> 43
>          'a Cluster_client.IDL.T.box = 'a Idl.Make(Idl.IdM).T.box
> 44
> make: *** [Makefile:13: build] Error 1
> 45
> Error: Process completed with exit code 2.```
> ```

Have you tried merging the latest xen-api#master?

**Comment by liulinC** [comment]
> > The unittest failed with IDL error, Looks like I need to merge some change in idl repo,
> > what should I do in this Koji world
> > ```
> > 40
> >          Cluster_interface.pems option ->
> > 41
> >          (unit, Cluster_interface.error) Cluster_client.IDL.T.resultb
> > 42
> >        but an expression was expected of type
> > 43
> >          'a Cluster_client.IDL.T.box = 'a Idl.Make(Idl.IdM).T.box
> > 44
> > make: *** [Makefile:13: build] Error 1
> > 45
> > Error: Process completed with exit code 2.```
> > ```
> 
> Have you tried merging the latest xen-api#master?

No, I was encouraged to only merge latest tag,  as last time I merge mater, the build got failure. 
I will try to merge latest master if encouraged.

**Comment by lippirk** [comment]
> No, I was encouraged to only merge latest tag, as last time I merge mater, the build got failure.
> I will try to merge latest master if encouraged.

I was recommending to pull in xen-api#master because it seems like you have somehow pulled in xcp-idl#master. You are right to use the latest tag rather than the master branch, as this _should_  just work. Seen as you're still having issues, I'd recommend you make sure you are not compiling against any of these commits https://github.com/xapi-project/xcp-idl/pull/331 and https://github.com/xapi-project/xen-api/pull/4458

**Comment by robhoes** [comment]
> somehow pulled in xcp-idl#master

The PR checks appear to use xcp-idl#master, probably because that comes from xs-opam. I had the same problem on my feature branch.

File: ocaml/xapi/extauth_plugin_ADwinbind.ml [hunk 2]
```diff
@@ -765,6 +780,29 @@ module Winbind = struct
...
 
+let build_netbios_name ~config_params =
+  let key = "netbios-name" in
+  match List.assoc_opt key config_params with
+  | Some name ->
+      if String.length name > max_netbios_name_length then
+        raise
+          (Auth_service_error
+             ( E_GENERIC
+             , Printf.sprintf "%s cannot longger than %d chars" key
```
Comment (lindig) [inline ocaml/xapi/extauth_plugin_ADwinbind.ml:791]: Grammar: "cannot be longer" or: "%s exceeds %d characters"

File: ocaml/xapi/extauth_plugin_ADwinbind.ml [hunk 3]
```diff
@@ -765,6 +780,29 @@ module Winbind = struct
...
+                 max_netbios_name_length ))
+      else
+        name
+  | None ->
+      Winbind.build_netbios_name ()
+
+let build_dns_hostname_option ~config_params =
```
Comment (lindig) [inline ocaml/xapi/extauth_plugin_ADwinbind.ml:798]: What does the "option" in the name signal here? This function does not return an option.

File: ocaml/xapi/extauth_plugin_ADwinbind.ml [hunk 4]
```diff
@@ -967,6 +1007,7 @@ module AuthADWinbind : Auth_signature.AUTH_MODULE = struct
       ; "--no-dns-updates"
       ]
       @ ou_param
+      @ dns_hostname_option
```
Comment (lindig) [inline ocaml/xapi/extauth_plugin_ADwinbind.ml:1010]: `@` is not efficient and you need to be careful about precedence and associativity. Better use `List..concat`

File: ocaml/xapi/extauth_plugin_ADwinbind.ml (hunk 1 above)
Comment (psafont) [reply ocaml/xapi/extauth_plugin_ADwinbind.ml:571]: Maybe the version of ocamlformat has changed, check with `ocamlformat -v` that 0.18.0 is being used

**Comment by liulinC** [comment]
Thanks @robhoes @lippirk @lindig @psafont 
I merged the master and the check got passed, 
However, it break my local build,  Can re-create a koji tag from rt-next from scratch resolve this issue?

File: ocaml/xapi/extauth_plugin_ADwinbind.ml [hunk 5]
```diff
@@ -72,6 +74,22 @@ let ntlm_auth uname passwd : (unit, exn) result =
...
+        h
+  in
+  Helpers.call_script ~log_output:On_failure net_cmd
+    ["lookup"; "kdc"; domain; "-d"; debug_level ()]
+  (* Result like 10.71.212.25:88\n10.62.1.25:88\n*)
+  |> String.split_on_char '\n'
+  |> hd "lookup kdc return invalid result"
+  |> String.split_on_char ':'
+  |> hd "kdc has invalid address"
+
```
Comment (lippirk) [inline ocaml/xapi/extauth_plugin_ADwinbind.ml:92]: I wonder if these IP addresses come in any particular order (e.g. sorted by round-trip time). I couldn't find any documentation on this

**Review by lippirk — APPROVED** [review]

File: ocaml/xapi/extauth_plugin_ADwinbind.ml (hunk 5 above)
Comment (liulinC) [reply ocaml/xapi/extauth_plugin_ADwinbind.ml:92]: Good question!
I did several times, and just found they comes randomly. 

I am currently working on another ticket to improve the ldap performance, basically, will query the latest kdc and persist in xapi database. I will raise PR when it is ready, 
For this PR, just keep it for now.

**Comment by liulinC** [comment]
@lindig  thanks for the comments, would you please help to review again, 
please just focus not the last commit, which has the corresponding refine and format.

File: ocaml/xapi/extauth_plugin_ADwinbind.ml [hunk 6]
```diff
@@ -765,6 +770,29 @@ module Winbind = struct
...
 
+let build_netbios_name ~config_params =
+  let key = "netbios-name" in
+  match List.assoc_opt key config_params with
+  | Some name ->
+      if String.length name > max_netbios_name_length then
+        raise
+          (Auth_service_error
+             ( E_GENERIC
+             , Printf.sprintf "%s cannot longger than %d chars" key
```
Comment (psafont) [inline ocaml/xapi/extauth_plugin_ADwinbind.ml]: typo: cannot longger -> be longer

File: ocaml/xapi/extauth_plugin_ADwinbind.ml [hunk 7]
```diff
@@ -201,6 +217,13 @@ module Ldap = struct
     let env = [|Printf.sprintf "KRB5_CONFIG=%s" domain_krb5_cfg|] in
     let* stdout =
       try
+        (* Query KDC instead of use domain here
+           * Just in case cannot resolve domain name from DSN *)
```
Comment (psafont) [inline ocaml/xapi/extauth_plugin_ADwinbind.ml:221]: Typo: DSN -> DNS

**Review by psafont — APPROVED** [review]
I'm not a domain expert, the changes in the code seem to match the behaviour changes explained in the commits

**Pull request merged**